### ISSUE TITLE
ready-only way of getting the nearest key to a non-existent key

### ISSUE THREAD

**Issue by chrbala** [Venemo/node-lmdb#116]
I am working on [lmdb-leveldown](https://github.com/chrbala/lmdb-leveldown) and am trying to get a reasonable way of getting the first key following/preceding a non-existent key. Right now, I am setting a dummy value to the key if it doesn't exist, then calling cursor.goToNext() or cursor.goToPrev() and aborting the transaction when done. However, this is cumbersome from an API standpoint and requires the use of a writable transaction. Is there a way to do this currently, or can it be added as a feature?

The broader use case for this is to support key range retrieval when the range keys may not be defined in the database.

**Comment by Venemo** [comment]
@chrbala Why not just use `goToRange`?

**Comment by chrbala** [comment]
Ah, I missed that in the examples! It looks like it mostly works, but not for keys outside of the keyspace. e.g. keys ["5", "6", "7"] are defined and we are trying to start at "9" and iterate backwards. It seems that goToRange only tests for keys after the one provided, so goToRange("9") here does not find a key.

**Comment by Venemo** [comment]
@chrbala `goToRange` will according to the LMDB docs *Position at first key greater than or equal to specified key.* So it does not have the exact functionality that you have in mind, but you can emulate it easily:

* Call goToRange`
* If the key is not found, call `goToPrev`.

Note that this does not require a write transaction.

FYI, node-lmdb is a wrapper around the lmdb C library, I recommend looking here for its docs: http://www.lmdb.tech/doc/

**Comment by chrbala** [comment]
Yeah, that's what I was doing, but it doesn't work when the key is out of range. It works correctly if you try goToRange("6") in ["4", "5", "7"]

```javascript
const goTo = (cursor: CursorType) => {
  const gt = key => {
    let curr = cursor.goToRange(key);
    while (curr && curr <= key) curr = cursor.goToNext();

    return curr;
  };

  const gte = key => {
    let curr = cursor.goToRange(key);
    while (curr && curr < key) curr = cursor.goToNext();

    return curr;
  };

  const lt = key => {
    let curr = cursor.goToRange(key);
    while (curr && curr >= key) curr = cursor.goToPrev();

    return curr;
  };

  const lte = key => {
    let curr = cursor.goToRange(key);
    while (curr && curr > key) curr = cursor.goToPrev();

    return curr;
  };
  return {
    gt,
    gte,
    lt,
    lte,
  };
};
```

From the looks of [the cursor docs](http://www.lmdb.tech/doc/group__mdb.html#gga1206b2af8b95e7f6b0ef6b28708c9127a2a35a3f4da5c7e8d24aad482c971def5), there may just not be a good way of supporting this due to a lack of support in the C library.

If you see no way of reasonably doing this, then the best path forward for me is likely filtering out the tests that do this and adding a note that this is breaking the API expectations from leveldown on the lmdb-leveldown docs.

Please resolve if you see no other path forward.

**Comment by Venemo** [comment]
@chrbala I'll look into it. I'm not familiar with that syntax though, is it Coffeescript?

**Comment by chrbala** [comment]
It's ES6 + Flow. You can see it transpiled to "vanilla" javascript on [the babel site](https://babeljs.io/repl/#?babili=false&browsers=&build=&builtIns=false&code_lz=MYewdgzgLgBA5iAKiGBeGAKYBXAThEXALhgGE8DdEBPABwFMBKNAPhgG8AoGGUSWOLHQBretVYduPGABt6sHLlxpeFQgDoEyAEoBDMHHoZR1RgG4pPAO4ALAJZzMi5QDIXqpTAA8Isc2cqipSaSCAAcvQAHlAY5pyWMLjyeGAeuBY8AL4WUnzQ8FD0KiYSXNKy8mmBarghOvqGxn4Z0rYORVh4ru4BXjAm_l3V-Bpa4VExcQlJUClpLdnxPHmwMkL9YqUJcgpD6EGjoXoGRgMt1vaOnZ5uVSy-plX7NXUgAApJAG6xOeUzc84Fr9eOB8msig8tuUdk8PMExsdGmcEm0rgFbgE2ANYQdamMPvRvlM_slcKlAVJFjx_mTJOVBAAaBKCehM6FQNnScGcxbZIA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&lineWrap=true&presets=es2015%2Creact%2Cstage-2&prettier=false&targets=&version=6.26.0&envVersion=)

You should be able to play around with my integration as shown above if you check out [my branch of lmdb-leveldown](https://github.com/chrbala/lmdb-leveldown/tree/chrbala)

**Comment by Venemo** [comment]
Ok, I will take a look at it as soon as I can.

**Comment by Venemo** [comment]
@chrbala Sorry for the delay. I thought about it now. Doesn't something like this satisfy your specification?

```js
cursor.goToNearest = function(searchKey) {
    var key = cursor.goToRange(searchKey);
    if (searchKey !== key) {
        key = cursor.goToPrev();
    }
    return key;
};
```

**Comment by Venemo** [comment]
@chrbala I just realized that I don't completely understand your requirement here, from the looks of your code maybe I misunderstood? Can you point me to a failing test suite where I could learn the desired behaviour and observe what is going on?

**Comment by chrbala** [comment]
Looks like I just overlooked the simple solution before:

``if (curr === null) curr = cursor.goToLast();``

I just released v0.0.3 of [lmdb-leveldown](https://www.npmjs.com/package/lmdb-leveldown), which should be the first stable release.

**Comment by Venemo** [comment]
@chrbala Awesome!
